File: src/main/java/com/clearcont/clearcontapp/model/ComposicaoLancamentosContabeis.java
package com.clearcont.clearcontapp.model;

import com.clearcont.clearcontapp.helpers.DecimalFormatBR;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.vaadin.flow.component.notification.Notification;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import org.hibernate.proxy.HibernateProxy;
import org.jetbrains.annotations.Nullable;

import java.text.NumberFormat;
import java.text.ParseException;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Locale;
import java.util.Objects;

@Entity
@AllArgsConstructor
@Getter
@Setter
@NoArgsConstructor
@Slf4j
public class ComposicaoLancamentosContabeis {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private LocalDate data;
    private String historico;
    @NotNull
    private Double debito = 0.0;
    @NotNull
    private Double credito = 0.0;
    private Double saldoContabil = debito - credito;
    @Enumerated(EnumType.STRING)
    private StatusConciliacao status = StatusConciliacao.OPEN;

    @Setter
    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.REMOVE)
    @JoinColumn(name = "balancete_id")
    private Balancete balancete;

    @Setter
    @ManyToOne
    private Responsavel responsavel;

    @OneToOne(fetch = FetchType.LAZY)    @JoinColumn(name = "customer_contabil_id")
    private CustomerContabil customerContabil;

    public ComposicaoLancamentosContabeis(Balancete balancete, Responsavel responsavel) {
        this.data = LocalDate.now();
        this.historico = "";
        this.debito = 0.0;
        this.credito = 0.0;
        this.saldoContabil = 0.0;
        this.status = StatusConciliacao.PROGRESS;
        this.balancete = balancete;
        this.responsavel = responsavel;
    }

    public ComposicaoLancamentosContabeis(Responsavel responsavel) {
        this.responsavel = responsavel;
    }

    public int contarPontos(@org.jetbrains.annotations.NotNull String texto) {
        int contador = 0;
        for (int i = 0; i < texto.length(); i++) {
            if (texto.charAt(i) == '.') {
                contador++;
            }
        }
        return contador;
    }

    public @org.jetbrains.annotations.NotNull String getDataFormated() {
        var formatador = DateTimeFormatter.ofPattern("dd/MM/yyyy").withLocale(Locale.of("pt", "BR"));
        return this.data.format(formatador);
    }


    public void setDebito(@org.jetbrains.annotations.NotNull String debito) {
        NumberFormat format = NumberFormat.getInstance(Locale.of("pt", "BR"));
        try {
            this.debito = format.parse(debito.replace("R$", "").trim()).doubleValue();
            this.saldoContabil = this.debito - this.credito;
        } catch (ParseException e) {
            log.info(e.getMessage());
            Notification.show("Erro");
        }
    }

    public void setCredito(@org.jetbrains.annotations.NotNull String credito) {
        NumberFormat format = NumberFormat.getInstance(Locale.of("pt", "BR"));
        try {
            this.credito = format.parse(credito.replace("R$", "").trim()).doubleValue();
            this.saldoContabil = this.debito - this.credito;
        } catch (ParseException e) {
            log.info(e.getMessage());
            Notification.show("Erro");
        }
    }

    public @org.jetbrains.annotations.NotNull String getDebito() {
        return DecimalFormatBR.getDecimalFormat().format(debito);
    }

    public @org.jetbrains.annotations.NotNull String getCredito() {
        return DecimalFormatBR.getDecimalFormat().format(credito);
    }

    public @org.jetbrains.annotations.NotNull String getSaldoContabil() {
        return DecimalFormatBR.getDecimalFormat().format(saldoContabil);
    }

    public double getDoubleSaldoContabil() {
        return saldoContabil;
    }

}

File: src/main/kotlin/br/com/clearcont/clearcontwebapp/models/ComposicaoLancamentosContabeis.kt
package br.com.clearcont.clearcontwebapp.models

import br.com.clearcont.clearcontwebapp.models.enums.StatusConciliacao
import br.com.clearcont.clearcontwebapp.models.enums.TipoConta
import com.fasterxml.jackson.annotation.JsonIgnore
import jakarta.persistence.*
import java.time.LocalDate
import java.time.Year
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.time.temporal.TemporalAdjusters
import java.util.*
import kotlin.jvm.Transient

@Entity
class ComposicaoLancamentosContabeis() : ComposicaoLancamentos() {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    var id: UUID? = null
    var numNotaFiscal = 0
    override var dataVencimento: LocalDate? = LocalDate.now()
    var ISS = 0.0
    var INSS = 0.0
    var IRRF = 0.0
    var CSRF = 0.0
    var debito: Double = 0.0
    var credito: Double = 0.0
    override var diasVencidos: Int = 0
        set(value) {
            field = if (value == 0) 1 else value
        }
    var historico: String? = ""

    @JsonIgnore
    @ManyToOne(
        fetch = FetchType.EAGER,
        cascade = [CascadeType.DETACH, CascadeType.REMOVE, CascadeType.PERSIST, CascadeType.REFRESH, CascadeType.MERGE]
    )
    @JoinColumn(name = "balancete_id")
    var balancete: Balancete? = null

    @Enumerated(EnumType.STRING)
    @Transient
    var status: StatusConciliacao? = null

    @ManyToOne
    lateinit var responsavel: Responsavel

    @PostLoad
    fun onLoad() {
        status = balancete?.status ?: StatusConciliacao.OPEN
        saldoContabil = debito - credito
    }

    init {
        calcularDiasVencidos(LocalDate.now().month.value)
    }

    private fun getFixedMonthValue(month: Int): Int = if (month in 1..12) month else 1

    override fun getDataVencimento(): String {
        val formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy")
        return dataVencimento!!.format(formatter)
    }

    override fun getDiasVencidos(month: Int): Int {
        calcularDiasVencidos(month)
        return diasVencidos
    }

    override fun calcularDiasVencidos(mes: Int) {
        val fixedMes = getFixedMonthValue(mes)
        val anoAtual = Year.now().value
        val ultimoDiaDoMes = LocalDate.of(anoAtual, fixedMes, 1).with(TemporalAdjusters.lastDayOfMonth())
        dataVencimento?.let {
            this.diasVencidos = ChronoUnit.DAYS.between(ultimoDiaDoMes, dataVencimento).toInt()

            if (diasVencidos < 0) {
                this.diasVencidos = -this.diasVencidos
            } else if (diasVencidos > 0) {
                this.diasVencidos = 0
            }
        }
    }

    override fun toString(): String {
        return "ComposicaoLancamentosContabeis(id=$id, numNotaFiscal=$numNotaFiscal, dataVencimento=$dataVencimento, ISS=$ISS, INSS=$INSS, IRRF=$IRRF, CSRF=$CSRF, debito=$debito, credito=$credito, diasVencidos=$diasVencidos, historico=$historico, balancete=$balancete, status=$status, responsavel=$responsavel)"
    }

    constructor(
        numNotaFiscal: Int,
        dataVencimento: LocalDate?,
        ISS: Double,
        INSS: Double,
        IRRF: Double,
        CSRF: Double
    ) : this() {
        this.numNotaFiscal = numNotaFiscal
        this.dataVencimento = dataVencimento
        this.ISS = ISS
        this.INSS = INSS
        this.IRRF = IRRF
        this.CSRF = CSRF
    }


    constructor(responsavel: Responsavel) : this() {
        this.responsavel = responsavel
    }

    constructor(
        id: UUID?,
        historico: String,
        debito: Double,
        credito: Double,
        balancete: Balancete?,
        responsavel: Responsavel,
        status: StatusConciliacao?
    ) : this() {
        this.id = id
        this.historico = historico
        this.debito = debito
        this.credito = credito
        this.balancete = balancete
        this.responsavel = responsavel
        this.status = status
    }

    constructor(
        numNotaFiscal: Int,
        dataVencimento: LocalDate?,
        data: LocalDate?,
        ISS: Double,
        INSS: Double,
        IRRF: Double,
        CSRF: Double,
        historico: String,
        credito: Double,
        debito: Double,
        balancete: Balancete?,
        status: StatusConciliacao?,
        responsavel: Responsavel
    ) : this() {
        super.data = data ?: LocalDate.now()
        this.numNotaFiscal = numNotaFiscal
        this.dataVencimento = dataVencimento
        this.ISS = ISS
        this.INSS = INSS
        this.IRRF = IRRF
        this.CSRF = CSRF
        this.historico = historico
        this.credito = credito
        this.debito = debito
        this.status = status
        this.balancete = balancete
        this.responsavel = responsavel
    }

    constructor(
        id: UUID?,
        numNotaFiscal: Int,
        dataVencimento: LocalDate?,
        data: LocalDate?,
        ISS: Double,
        INSS: Double,
        IRRF: Double,
        CSRF: Double,
        diasVencidos: Int,
        historico: String,
        debito: Double,
        credito: Double,
        balancete: Balancete?,
        status: StatusConciliacao?,
        responsavel: Responsavel
    ) : this() {
        this.id = id
        this.numNotaFiscal = numNotaFiscal
        this.dataVencimento = dataVencimento
        super.data = data ?: LocalDate.now()
        this.ISS = ISS
        this.INSS = INSS
        this.IRRF = IRRF
        this.CSRF = CSRF
        this.diasVencidos = diasVencidos
        this.historico = historico
        this.debito = debito
        this.credito = credito
        this.balancete = balancete
        this.status = status
        this.responsavel = responsavel
    }

    constructor(balancete: Balancete?, responsavel: Responsavel?) : this() {
        this.balancete = balancete
        this.responsavel = responsavel!!
    }



}
